### PR TITLE
Add distributed ndarray

### PR DESCRIPTION
**Pull request by shino16**
Adds `array.DistributedArray` to `cupyx.distributed`.

It provides initial support of

* conversion from/to `ndarray`
* element-wise operations (`ufunc`, `ElementwiseKernel`)
* reduction (max/min/sum/prod)
* matrix multiplication (`matmul`)

in a multi-GPU setting.

### CODE REVIEW

File: cupy/_core/_reduction.pyx [hunk 1]
```diff
@@ -581,6 +581,10 @@ cdef class _SimpleReductionKernel(_AbstractReductionKernel):
     def __call__(self, object a, axis=None, dtype=None, _ndarray_base out=None,
                  bint keepdims=False):
 
+        if hasattr(a, '__cupy_override_reduction_kernel__'):
```
Comment (emcastillo) [inline cupy/_core/_reduction.pyx:584]: This is currently only working for `SimpleReductionKernel`
There are other more generic Reduction kernels, I wonder if we can easily support them?

File: cupy/_core/core.pyx [hunk 2]
```diff
@@ -995,6 +995,9 @@ cdef class _ndarray_base:
            :meth:`numpy.ndarray.max`
 
         """
+        if hasattr(self, '__cupy_override_reduction_kernel__'):
```
Comment (emcastillo) [inline cupy/_core/core.pyx:998]: I would only like to have this check in the `ReductionKernel` machinery. Do you think it should be possible to do that? thanks!

File: cupyx/distributed/_nccl_comm.py [hunk 3]
```diff
@@ -43,6 +43,18 @@
     _nccl_ops = {}
 
 
+def _get_nccl_dtype_and_count(array, count=None):
```
Comment (emcastillo) [inline cupyx/distributed/_nccl_comm.py:46]: nice!

File: cupyx/distributed/array/_array.py [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+        """Instantiate a distributed array using arguments for its attributes.
+
+        :func:`~cupy.distributed.array.distributed_array` provides a more user
+        friendly way of creating a distributed array from another array.
+
+        Args:
+            shape (tuple of ints): Length of axes.
+            dtype: Data type. It must be an argument of :class:`numpy.dtype`.
+            mode (str or mode object): Mode that determines how overlaps of
+                chunks are interpreted.
```
Comment (emcastillo) [inline cupyx/distributed/array/_array.py:67]: I would like a description of the available models in the docstring under a `.. note:` section :)

File: cupy/_core/_reduction.pyx (hunk 1 above)
Comment (shino16) [reply cupy/_core/_reduction.pyx:584]: `__cupy_override_reduction_kernel__` has to be called here, before type checks which rule out `DistributedArray`.
I believe we can support `ReductionKernel` in the same way, by adding this hook in its `__call__` method.

File: cupy/_core/core.pyx (hunk 2 above)
Comment (shino16) [reply cupy/_core/core.pyx:998]: If we remove this check, `cupy._core._routines_statistics._ndarray_max` tries CUB and cuTENSOR before falling back to `cupy.max`.
So this check should be done before reaching `cupy.max.__call__`, I wonder where this check should ideally be placed.
https://github.com/shino16/cupy/blob/main/cupy/_core/_routines_statistics.pyx#L25-L43

Comment (emcastillo) [reply cupy/_core/core.pyx:998]: That's a good call. Maybe we should rewrite those checks to make it feasible.
Like having an attribute in the ndarray that's overridden by derived classes.
`array.support_cub_for_routines`

Comment (shino16) [reply cupy/_core/core.pyx]: Now `DistributedArray` overrides all the methods of `ndarray` including max/min/sum/prod, so those hooks in `cupy._core.core.ndarray` are removed. Still your idea sounds reasonable.

File: examples/distributed/elementwise.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+INDEX_MAP_A = {
+    0: (slice(200)),  # arr[:200]
+    1: (slice(200, None)),  # arr[200:]
+}
+
+INDEX_MAP_B = {
+    0: (slice(None), slice(None, None, 3)),  # arr[:,  ::2]
+    1: (slice(None), slice(1, None, 2)),  # arr[:, 1::2]
+}
```
Comment (leofang) [inline examples/distributed/elementwise.py:19]: It'd be nice to comment why we need these index maps, currently this sample assumes that users have the concept of sharding, which may or may not be the case.

File: examples/distributed/elementwise.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+b = distributed_array(data_b, INDEX_MAP_B)
+
+with Device(0):
+    s0 = Stream()
+    s0.use()
+
+with Device(1):
+    s1 = Stream()
+    s1.use()
```
Comment (leofang) [inline examples/distributed/elementwise.py:34]: Quality of life change: This sample assumes there exists 2 GPUs, we need a device count check at the top and do early exit.

File: examples/distributed/matmul.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+data_a = numpy.arange(N * M).reshape(N, M)
+data_b = numpy.arange(M * P).reshape(M, P)
+
+#     0    M/3    M
+#   0 +-----+-----+
+#     | 0 1 |  2  |
+# N/2 +-----+-----+
+#     | 0 3 | 1 2 |
+#   N +-----+-----+
```
Comment (leofang) [inline examples/distributed/matmul.py:19]: Q: Does this sample require 4 GPUs?

**Review by leofang — COMMENTED** [review]
Quick drive-by comments 🙂

File: examples/distributed/matmul.py (hunk 7 above)
Comment (emcastillo) [reply examples/distributed/matmul.py:19]: Yeah, we need to add a runtime check here! great catch :)

File: cupyx/distributed/array/_array.py [hunk 8]
```diff
@@ -0,0 +1,917 @@
...
+    __init__(self, shape, dtype, chunks_map, mode=_REPLICA_MODE, comms=None)
+
+    Multi-dimensional array distributed across multiple CUDA devices.
+
+    This class implements some elementary operations that :class:`cupy.ndarray`
+    provides. The array content is split into chunks, contiguous arrays
+    corresponding to slices of the original array. Note that one device can
+    hold multiple chunks.
+
+    This direct constructor is designed for internal calls. Users should create
```
Comment (emcastillo) [inline cupyx/distributed/array/_array.py:51]:
```suggestion
    This direct constructor is intended as a private API. Users should create
```

File: cupyx/distributed/array/_array.py [hunk 9]
```diff
@@ -0,0 +1,917 @@
...
+        if isinstance(mode, str):
+            if mode not in _modes._MODES.keys():
+                raise RuntimeError(
+                    f'`mode` must be one of {list(_modes._MODES)}')
+            mode = _modes._MODES[mode]
+        obj._mode = mode
+
+        obj._streams = {}
+        obj._comms = comms if comms is not None else {}
+
```
Comment (emcastillo) [inline cupyx/distributed/array/_array.py:95]: Several of the assignments here could be moved to __init__

File: cupyx/distributed/array/_array.py [hunk 10]
```diff
@@ -0,0 +1,917 @@
...
+    """
+    if isinstance(array, DistributedArray):
+        if array.mode != mode:
+            array = array.change_mode(mode)
+        if array.index_map != index_map:
+            array = array.reshard(index_map)
+        return DistributedArray(
+            array.shape, array.dtype, array._chunks_map, array._mode,
+            array._comms)
+
```
Comment (emcastillo) [inline cupyx/distributed/array/_array.py:864]: TODO
Support regular CuPy arrays and distribute the chunks without needing to synchronize

File: cupyx/distributed/array/_array.py [hunk 11]
```diff
@@ -0,0 +1,917 @@
...
+    def strides(self):
+        """Not supported."""
+        raise NotImplementedError(
+            'DistributedArray currently does not support strides.')
+
+
+def distributed_array(
+    array: ArrayLike,
+    index_map: dict[int, Any],
+    mode: str = 'replica',
```
Comment (emcastillo) [inline cupyx/distributed/array/_array.py:828]: maybe better to use enum?

File: cupyx/distributed/array/_array.py [hunk 12]
```diff
@@ -0,0 +1,917 @@
...
+            array = array.change_mode(mode)
+        if array.index_map != index_map:
+            array = array.reshard(index_map)
+        return DistributedArray(
+            array.shape, array.dtype, array._chunks_map, array._mode,
+            array._comms)
+
+    if isinstance(array, (numpy.ndarray, ndarray)):
+        if mode != 'replica':
+            array = array.copy()
```
Comment (emcastillo) [inline cupyx/distributed/array/_array.py:867]: TODO: If the array is already in the device, making a copy can cause an OOM, in some cases we may be able to copy only the chunk in the device if the array is contiguous.

File: cupyx/distributed/array/_chunk.py [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+from cupy.cuda.stream import get_current_stream
+
+from cupyx.distributed.array import _modes
+from cupyx.distributed.array import _index_arith
+from cupyx.distributed.array import _data_transfer
+from cupyx.distributed.array._data_transfer import _Communicator
+
+
+class _ArrayPlaceholder:
+    # Mocks ndarray
```
Comment (emcastillo) [inline cupyx/distributed/array/_chunk.py:22]: Self note, this is used when we are waiting chunks to arrive, to the current device

**Comment by emcastillo** [comment]
@shino16 can you repush your changes and open again the PR? thanks!

**Comment by kmaehashi** [comment]
This pull-request has merged as #7942. Thanks again @shino16 for working on this!

**Pull request closed without merging**